source.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2017 Pilosa Corp.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+// CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+
+package pdk
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+// PeekingSource is a wrapper for Source which implements the
+// SourcePeeker interface by reading the next record from Source
+// and caching it for the next call to Record().
+type PeekingSource struct {
+	Source
+
+	mu       sync.RWMutex
+	rec      interface{}
+	recordFn func() (interface{}, error)
+}
+
+func (p *PeekingSource) Peek() (interface{}, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	var err error
+	p.rec, err = p.Source.Record()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting next record for peeking")
+	}
+	p.recordFn = p.cachedRecord
+	return p.rec, nil
+}
+
+func (p *PeekingSource) cachedRecord() (interface{}, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.recordFn = p.Source.Record
+	return p.rec, nil
+}
+
+func (p *PeekingSource) Record() (interface{}, error) {
+	if p.recordFn == nil {
```

Comment:
I don't think this is threadsafe. p.recordFn could be read here while written in cachedRecord. maybe we should declare PeekingSource to be not threadsafe and forgo the locking and recordFn complexity?

Reply:
Yeah, this is definitely not threadsafe. I strugged with this part of the code a bit. I'm trying to prevent the need for locking within the `Record()` method, since presumably that's the one that needs to support high volume. 

Would it be threadsafe if `Record()` was just this:
```
func (p *PeekingSource) Record() (interface{}, error) {
    return p.recordFn()
}
```
In order to make that work, we would need to have a creator function that returns a `*PeekingSource` that sets up `p.recordFn = p.Source.Record` as default on creation.

Reply:
Are you saying in that last sentence that only one goroutine would have access to a given instance of a PeekingSource? if not it still would not be thread safe

Reply:
@tgruben so are the only options to either:
1. introduce a lock in `Record()`
or
2. use a channel
?

is one better than the other?